From 5159ac695da65ebd4136295686149ccc70a15f1c Mon Sep 17 00:00:00 2001
From: Wenzong Fan <wenzong.fan@windriver.com>
Date: Fri, 2 Feb 2018 09:31:35 +0000
Subject: [PATCH] wr-policy: fix labels for resolv.conf

Port fixes from https://github.com/fedora-selinux/selinux-policy:
  b84a0034 Add label for /var/run/systemd/resolve/resolv.conf.
  ccc999d0 Create resolv.conf in the pppd_var_run_t with the \
           net_conf_t label
  9fe14b77 Make systemd resolved to create resolv.conf with \
           correct label.

  Pick changes from:
  cc28df82 Fix labels, improve sysnet_manage_config interface.
  2fe649cd Add support systemd-resolved.

WRL Fixes:
  * Allow dhclient to create /tmp/resolv.conf.dhclient-new with type
    net_conf_t;
  * Allow dhclient to remove symlink /etc/resolv.conf;
  * Replace /var/run/systemd with /run/systemd to fix default labels
    for /var/run/systemd/resolve/resolv.conf;
  * Set default labels(net_conf_t) for the symlinks:
    /etc/resolv.conf
    /etc/resolv-conf.systemd

How resolv.conf is created:
  * dhclient creates /tmp/resolv.conf.dhclient-new and then move as
    /etc/resolv.conf;
  * systemd-resolved creates /var/run/systemd/resolve/.#resolv.confX3Arq8"
    (random name) and then move as /var/run/systemd/resolve/resolv.conf;
  * systemd creates symlinks:
    /etc/resolv.conf -> /etc/resolv-conf.systemd
    /etc/resolv-conf.systemd -> /var/run/systemd/resolve/resolv.conf

Remove avc denial like:
  avc: denied { search } for pid=661 comm="dhclient-script" \
  name="resolve" dev="tmpfs" ino=16673 \
  scontext=system_u:system_r:dhcpc_t \
  tcontext=system_u:object_r:systemd_resolved_var_run_t tclass=dir

Upstream-Status: Pending

Signed-off-by: Wenzong Fan <wenzong.fan@windriver.com>
---
 policy/modules/system/sysnetwork.fc |  5 ++++-
 policy/modules/system/sysnetwork.if | 35 +++++++++++++++++++++++++++++++++++
 policy/modules/system/sysnetwork.te |  1 +
 policy/modules/system/systemd.te    |  3 +++
 4 files changed, 43 insertions(+), 1 deletion(-)

diff --git a/policy/modules/system/sysnetwork.fc b/policy/modules/system/sysnetwork.fc
index 52f52b0..c0bcc0a 100644
--- a/policy/modules/system/sysnetwork.fc
+++ b/policy/modules/system/sysnetwork.fc
@@ -26,7 +26,7 @@ ifdef(`distro_debian',`
 /etc/hosts\.deny.*	--	gen_context(system_u:object_r:net_conf_t,s0)
 /etc/denyhosts.*	--	gen_context(system_u:object_r:net_conf_t,s0)
 /etc/resolv\.conf.*	--	gen_context(system_u:object_r:net_conf_t,s0)
-/var/run/resolv\.conf.*	--	gen_context(system_u:object_r:net_conf_t,s0)
+/run/resolv\.conf.*	--	gen_context(system_u:object_r:net_conf_t,s0)
 /etc/yp\.conf.*		--	gen_context(system_u:object_r:net_conf_t,s0)
 
 /etc/dhcp3(/.*)?		gen_context(system_u:object_r:dhcp_etc_t,s0)
@@ -36,6 +36,9 @@ ifdef(`distro_redhat',`
 /etc/sysconfig/network-scripts/.*resolv\.conf -- gen_context(system_u:object_r:net_conf_t,s0)
 /etc/sysconfig/networking(/.*)? gen_context(system_u:object_r:net_conf_t,s0)
 /etc/sysconfig/network-scripts(/.*)? gen_context(system_u:object_r:net_conf_t,s0)
+/run/systemd/resolve/resolv\.conf   --  gen_context(system_u:object_r:net_conf_t,s0)
+/etc/resolv-conf\.systemd	-l	gen_context(system_u:object_r:net_conf_t,s0)
+/etc/resolv\.conf	-l	gen_context(system_u:object_r:net_conf_t,s0)
 ')
 
 #
diff --git a/policy/modules/system/sysnetwork.if b/policy/modules/system/sysnetwork.if
index 7e16b2b..bf89a0b 100644
--- a/policy/modules/system/sysnetwork.if
+++ b/policy/modules/system/sysnetwork.if
@@ -441,6 +441,40 @@ interface(`sysnet_etc_filetrans_config',`
 	files_etc_filetrans($1, net_conf_t, file, $2)
 ')
 
+########################################
+## <summary>
+##	Transition content to the type used for
+##	the network config files.
+## </summary>
+## <param name="domain">
+##	<summary>
+##      Domain allowed access.
+##	</summary>
+## </param>
+## <param name="private type">
+##	<summary>
+##	The type of the directory to which the object will be created.
+##	</summary>
+## </param>
+## <param name="object_class">
+##	<summary>
+##	The object class.
+##	</summary>
+## </param>
+## <param name="name" optional="true">
+##	<summary>
+##	The name of the object being created.
+##	</summary>
+## </param>
+#
+interface(`sysnet_filetrans_config_fromdir',`
+	gen_require(`
+		type net_conf_t;
+	')
+
+	filetrans_pattern($1, $2, net_conf_t, $3, $4)
+')
+
 #######################################
 ## <summary>
 ##	Create, read, write, and delete network config files.
@@ -466,6 +500,7 @@ interface(`sysnet_manage_config',`
 	ifdef(`distro_redhat',`
 		files_search_all_pids($1)
 		manage_files_pattern($1, net_conf_t, net_conf_t)
+		manage_lnk_files_pattern($1, net_conf_t, net_conf_t)
 	')
 ')
 
diff --git a/policy/modules/system/sysnetwork.te b/policy/modules/system/sysnetwork.te
index 8f717cd..59e02e6 100644
--- a/policy/modules/system/sysnetwork.te
+++ b/policy/modules/system/sysnetwork.te
@@ -78,6 +78,7 @@ filetrans_pattern(dhcpc_t, var_run_t, net_conf_t, file, "resolv.conf.dhclient6")
 # in /etc created by dhcpcd will be labelled net_conf_t.
 sysnet_manage_config(dhcpc_t)
 files_etc_filetrans(dhcpc_t, net_conf_t, file)
+files_tmp_filetrans(dhcpc_t, net_conf_t, file, "resolv.conf.dhclient-new")
 
 # create temp files
 manage_dirs_pattern(dhcpc_t, dhcpc_tmp_t, dhcpc_tmp_t)
diff --git a/policy/modules/system/systemd.te b/policy/modules/system/systemd.te
index d73f440..a685e56 100644
--- a/policy/modules/system/systemd.te
+++ b/policy/modules/system/systemd.te
@@ -379,6 +379,9 @@ corenet_udp_bind_llmnr_port(systemd_resolved_t)
 corenet_udp_bind_dns_port(systemd_resolved_t)
 corenet_tcp_bind_dns_port(systemd_resolved_t)
 
+sysnet_manage_config(systemd_resolved_t)
+sysnet_filetrans_config_fromdir(systemd_resolved_t,systemd_resolved_var_run_t, file)
+
 auth_use_nsswitch(systemd_resolved_t)
 
 seutil_read_file_contexts(systemd_resolved_t)
-- 
2.13.3

